File: pkg/store/draft_store.go
package store

import (
	"context"
	"errors"
	"fmt"

	"gorm.io/gorm"
	"gorm.io/gorm/clause"

	"github.com/PineappleBond/xyncra-server/pkg/store/model"
)

// DraftStore provides data access operations for message drafts.
// Each conversation can have at most one draft (one-draft-per-conversation).
type DraftStore struct {
	db *gorm.DB
}

// Save performs an UPSERT for a draft. If a draft for the conversation already
// exists (by ConversationID uniqueIndex), it is updated; otherwise a new
// record is inserted.
func (ds *DraftStore) Save(ctx context.Context, draft *model.Draft) error {
	err := ds.db.WithContext(ctx).
		Clauses(clause.OnConflict{
			Columns:   []clause.Column{{Name: "conversation_id"}},
			DoUpdates: clause.AssignmentColumns([]string{"content", "updated_at"}),
		}).
		Create(draft).Error
	if err != nil {
		return classifyError(fmt.Errorf("store: save draft: %w", err))
	}
	return nil
}

// GetByConversation retrieves the draft for the given conversation.
// Returns ErrNotFound if no draft exists.
func (ds *DraftStore) GetByConversation(ctx context.Context, conversationID string) (*model.Draft, error) {
	var draft model.Draft
	err := ds.db.WithContext(ctx).
		Where("conversation_id = ?", conversationID).
		First(&draft).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, ErrNotFound
		}
		return nil, classifyError(fmt.Errorf("store: get draft: %w", err))
	}
	return &draft, nil
}

// Delete removes a draft by its primary key. Returns ErrNotFound if not found.
func (ds *DraftStore) Delete(ctx context.Context, id string) error {
	result := ds.db.WithContext(ctx).Delete(&model.Draft{}, "id = ?", id)
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: delete draft: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	return nil
}

// DeleteByConversation removes the draft for the given conversation.
// Returns ErrNotFound if no draft exists.
func (ds *DraftStore) DeleteByConversation(ctx context.Context, conversationID string) error {
	result := ds.db.WithContext(ctx).
		Where("conversation_id = ?", conversationID).
		Delete(&model.Draft{})
	if result.Error != nil {
		return classifyError(fmt.Errorf("store: delete draft by conversation: %w", result.Error))
	}
	if result.RowsAffected == 0 {
		return ErrNotFound
	}
	return nil
}

// List returns all drafts ordered by UpdatedAt descending.
func (ds *DraftStore) List(ctx context.Context) ([]*model.Draft, error) {
	var drafts []*model.Draft
	err := ds.db.WithContext(ctx).
		Order("updated_at DESC").
		Find(&drafts).Error
	if err != nil {
		return nil, classifyError(fmt.Errorf("store: list drafts: %w", err))
	}
	return drafts, nil
}

File: demo/web/packages/xyncra-client-core/src/db/draft-store.ts
/**
 * DraftStore — data access operations for message drafts.
 *
 * Mirrors Go DraftStore (pkg/store/draft_store.go).
 *
 * Each conversation can have at most one draft (conversation_id is uniqueIndex).
 */

import { ErrNotFound } from '../errors';
import type { XyncraDatabase } from './index';
import type { Draft } from './models';

/**
 * DraftStore provides data access operations for message drafts.
 * Each conversation can have at most one draft (one-draft-per-conversation).
 */
export class DraftStore {
  constructor(private readonly db: XyncraDatabase) {}

  /**
   * Performs an UPSERT for a draft. If a draft for the conversation already
   * exists (by conversation_id uniqueIndex), it is updated; otherwise a new
   * record is inserted.
   */
  async save(draft: Draft): Promise<void> {
    // Check for existing draft by conversation_id.
    const existing = await this.db.drafts
      .where('conversation_id')
      .equals(draft.conversation_id)
      .first();

    if (existing) {
      // Update existing draft.
      await this.db.drafts.update(existing.id, {
        content: draft.content,
        updated_at: new Date(),
      });
    } else {
      await this.db.drafts.add(draft);
    }
  }

  /**
   * Retrieves the draft for the given conversation.
   * Throws ErrNotFound if no draft exists.
   */
  async getByConversation(conversationID: string): Promise<Draft> {
    const draft = await this.db.drafts
      .where('conversation_id')
      .equals(conversationID)
      .first();
    if (!draft) {
      throw ErrNotFound;
    }
    return draft;
  }

  /**
   * Removes a draft by its primary key.
   * Throws ErrNotFound if not found.
   */
  async delete(id: string): Promise<void> {
    const existing = await this.db.drafts.get(id);
    if (!existing) {
      throw ErrNotFound;
    }
    await this.db.drafts.delete(id);
  }

  /**
   * Removes the draft for the given conversation.
   * Throws ErrNotFound if no draft exists.
   */
  async deleteByConversation(conversationID: string): Promise<void> {
    const draft = await this.db.drafts
      .where('conversation_id')
      .equals(conversationID)
      .first();
    if (!draft) {
      throw ErrNotFound;
    }
    await this.db.drafts.delete(draft.id);
  }

  /**
   * Returns all drafts ordered by updated_at descending.
   */
  async list(): Promise<Draft[]> {
    const drafts = await this.db.drafts.toArray();
    drafts.sort(
      (a, b) =>
        new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime(),
    );
    return drafts;
  }
}
